pkg/image/registryclient/client_mirrored.go

```diff
@@ -0,0 +1,477 @@
+type mirroredBlobstore struct {
+	repo *mirroredRepository
+}
+
+var _ distribution.BlobService = mirroredBlobstore{}
+
+func (f mirroredBlobstore) Get(ctx context.Context, dgst digest.Digest) ([]byte, error) {
+	var data []byte
+	err := f.repo.alternates(ctx, func(r distribution.Repository) error {
+		var err error
+		data, err = r.Blobs(ctx).Get(ctx, dgst)
+		klog.V(5).Infof("get for %s served from %s: %v", dgst, r.Named(), err)
+		return err
+	})
+	return data, err
+}
+
+func (f mirroredBlobstore) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
+	var desc distribution.Descriptor
+	err := f.repo.alternates(ctx, func(r distribution.Repository) error {
+		var err error
+		desc, err = r.Blobs(ctx).Stat(ctx, dgst)
+		return err
+	})
+	return desc, err
+}
+
+func (f mirroredBlobstore) ServeBlob(ctx context.Context, w http.ResponseWriter, req *http.Request, dgst digest.Digest) error {
+	err := f.repo.firstConnectedAlternate(ctx, func(r distribution.Repository) error {
+		return r.Blobs(ctx).ServeBlob(ctx, w, req, dgst)
+	})
+	return err
+}
+
+func (f mirroredBlobstore) Open(ctx context.Context, dgst digest.Digest) (distribution.ReadSeekCloser, error) {
+	var rsc distribution.ReadSeekCloser
+	err := f.repo.alternates(ctx, func(r distribution.Repository) error {
+		var err error
+		rsc, err = r.Blobs(ctx).Open(ctx, dgst)
+		return err
+	})
+	return rsc, err
+}
+
+func (f mirroredBlobstore) Create(ctx context.Context, options ...distribution.BlobCreateOption) (distribution.BlobWriter, error) {
+	var bw distribution.BlobWriter
+	err := f.repo.source(ctx, func(r distribution.Repository) error {
+		var err error
+		bw, err = r.Blobs(ctx).Create(ctx, options...)
+		return err
+	})
+	return bw, err
+}
+
+func (f mirroredBlobstore) Put(ctx context.Context, mediaType string, p []byte) (distribution.Descriptor, error) {
+	var desc distribution.Descriptor
+	err := f.repo.source(ctx, func(r distribution.Repository) error {
+		var err error
+		desc, err = r.Blobs(ctx).Put(ctx, mediaType, p)
+		return err
+	})
+	return desc, err
+}
+
+func (f mirroredBlobstore) Resume(ctx context.Context, id string) (distribution.BlobWriter, error) {
+	var bw distribution.BlobWriter
+	err := f.repo.source(ctx, func(r distribution.Repository) error {
+		var err error
+		bw, err = r.Blobs(ctx).Resume(ctx, id)
+		return err
+	})
+	return bw, err
+}
+
+func (f mirroredBlobstore) Delete(ctx context.Context, dgst digest.Digest) error {
+	return f.repo.source(ctx, func(r distribution.Repository) error {
+		return r.Blobs(ctx).Delete(ctx, dgst)
+	})
+}
+
+// mirroredTags lazily accesses the source repository
+type mirroredTags struct {
+	repo *mirroredRepository
+}
+
+var _ distribution.TagService = mirroredTags{}
+
+func (f mirroredTags) Get(ctx context.Context, tag string) (distribution.Descriptor, error) {
+	var desc distribution.Descriptor
+	err := f.repo.source(ctx, func(r distribution.Repository) error {
```

Comment:
no because you don't know whether they are in sync.  I think that would be a different type of behavior - we could call this "blobMirrored" to clarify because everything has to be content addressable.

Reply:
Renamed to make it a bit clearer and updated the godoc on the top most item.

Reply:
i've been stuck w/ the repository going from mirrored(blobMirroredRepository) to retryRepository w/ client.go `connectToRegistry` and when I try to mirror any image I see a manifestUnknown w/ the manifests Put  - If I don't fix this I'll push my oc branch and have you try it/help tomorrow. (first I'll incorporate the name changes above)

Reply:
ooooh got it to work - `oc` was considering a nil manifest and an empty Descriptor as 'already existing' once I fixed that the push goes through.. now, why is the regclient returning a nil manifest and empty Descriptor?